Add ActiveRecord::Base.destroy_later

georgeclaghorn opened the pull request:
Generalize Action Mailbox's incineration feature, as proposed by @dhh in rails/actionmailbox#16. Permit automatically destroying records after a specified amount of time:

```ruby
class Export < ApplicationRecord
  # Destroy all exports 30 days after creation
  destroy_later after: 30.days

  # Destroy exports 30 days after completing them
  destroy_later after: 30.days, if: -> { status_previously_changed? && completed? }

  # Destroy exports 30 days after completing them.
  # Ensure they're still completed at destroy time.
  destroy_later after: 30.days,
    if: -> { status_previously_changed? && completed? }, ensuring: :completed?
end
```

The implementation as it stands now is not final. I expect to find a way to avoid the hard dependency on Active Job—it will only be required to use `destroy_later`.

hahmed commented:
@georgeclaghorn was reviewing this code, and like where this is going. We are adding code into AR for destroying records later, there is also code related to `actionmailbox` in this PR too, would it be worth splitting this out?

georgeclaghorn committed:
Add ActiveRecord::Base.destroy_later

Generalize Action Mailbox's incineration feature. Permit destroying records after a specified amount of time.

georgeclaghorn commented:
That’s a fair point, @hahmed. I removed the Action Mailbox changes from this PR and will apply them separately.

matthewd commented on activerecord/lib/active_record/destroy_later.rb:
```diff
@@ -0,0 +1,38 @@
+
+    module ClassMethods
+      # :call-seq: destroy_later(after:, if: nil, ensuring: nil)
+      def destroy_later(after:, ensuring: nil, **options)
```
I'm not sure how we (or even I) feel about one vs the other, but FWIW I believe the officially-recommended way to do this is:

```ruby
def destroy_later(after:, if: nil, ensuring: nil)
  if_ = binding.local_variable_get(:if)
```

georgeclaghorn replied:
I like that this results in an `ArgumentError` if you pass an unsupported option. I went without the extra local variable.

kaspth commented on activerecord/lib/active_record/destroy_job.rb:
```diff
@@ -0,0 +1,18 @@
+
+    private
+      def eligible?(record, ensuring)
+        ensuring.nil? || record.public_send(ensuring)
```
What should we do in case someone misspells `ensuring` or has renamed it later and forgotten that it was used for `destroy_later`? Too arcane to bother with, should we catch the `NoMethodError`, something else?

georgeclaghorn replied:
I think letting the `NoMethodError` bubble up to the job runner is legit.

In the first case, where the method name in `ensuring:` is incorrect or misspelled, you’ll end up with a list of jobs to correct in your Active Job backend’s failed queue. In the second case, you’ll want to restore the method and retry the failed jobs.

georgeclaghorn commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -31,6 +32,8 @@ class Railtie < Rails::Railtie # :nodoc:
     config.active_record.sqlite3 = ActiveSupport::OrderedOptions.new
     config.active_record.sqlite3.represent_boolean_as_integer = nil
 
+    config.active_record.queues = ActiveSupport::InheritableOptions.new(destroy: :active_record_destroy)
```
More than a few. ;)

We like sending built-in jobs to separate queues by default because it allows us to prioritize them. For example, we can configure a pool of Resque workers to serve jobs in the following prioritized order:

1. `chat` (In-app jobs that write new chat lines to the DB; most sensitive to delay)
1. `action_mailbox_routing`
2. `active_storage_analysis`
3. `search` (In-app jobs that update Elasticsearch indexes)
4. `active_record_destroy`
5. `active_storage_purge` (Least sensitive to delay)

Using separate queues also allows us to stop work on a particular queue when the jobs it hosts put too much pressure on the DB or an external service.

The guides should probably list the default queues for built-in jobs and explicitly state that it may be necessary to configure the Active Job backend to serve them. I also want to update the API docs for each of the built-in jobs to mention their respective default queues.

kaspth replied:
All good reasons, thanks! Agreed with the guides, might be nice to add the advice about serving queues in prioritized order as well.

georgeclaghorn commented on activerecord/lib/active_record/destroy_later.rb:
```diff
@@ -0,0 +1,38 @@
+      # :call-seq: destroy_later(after:, if: nil, ensuring: nil)
+      def destroy_later(after:, ensuring: nil, **options)
+        after_commit -> { destroy_later after: after, ensuring: ensuring },
+          on: options[:if] ? %i[ create update ] : :create, if: options[:if]
```
I’d prefer not to. We don’t have a use case for it in Basecamp, and I think it’d need to be paired with an inverse for `ensuring` to be useful (which complicates the feature and its implementation significantly).

georgeclaghorn commented on activerecord/lib/active_record/destroy_job.rb:
```diff
@@ -0,0 +1,18 @@
+    end
+
+    private
+      def eligible?(record, ensuring)
```
That’ll be up to models to implement in their `ensuring` methods if they need it.

georgeclaghorn replied:
I think we do! I’ll restore the due check in the follow-up PR that replaces Action Mailbox incineration with `destroy_later`.

kaspth replied:
Got it. Is this something we should extract as well? I remember seeing that pattern everywhere in one of David's screencasts. Maybe something like `require_untouched: true` and perhaps `require_untouched: :created_at`.

georgeclaghorn commented on activerecord/lib/active_record/destroy_later.rb:
```diff
@@ -0,0 +1,38 @@
+  #
+  #   # Destroy exports 30 days after completing them
+  #   class Export < ApplicationRecord
+  #     destroy_later after: 30.days, if: -> { completed_previously_changed? && completed? }
```
Added:

> `destroy_later` adds an `after_commit` callback that schedules an ActiveRecord::DestroyJob. In the example above, we use `completed_previously_changed?` to check whether the record's `completed` flag was changed in the preceding save. (`completed_changed?` wouldn't work because any prior changes to the record have been saved at the time the condition is evaluated.)

georgeclaghorn replied:
I don’t see a great place in the API docs for `ActiveModel::Dirty` or any of the guides.

Ideally, something would explain that the `*_changed?` and `*_previously_changed?` methods are added automatically for each attribute and clearly state the difference between them. The `ActiveModel::Dirty` docs don’t do that; they’re better suited for people who are mixing it into their own (non–Active Record) classes.

kaspth commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -31,6 +32,8 @@ class Railtie < Rails::Railtie # :nodoc:
     config.active_record.sqlite3 = ActiveSupport::OrderedOptions.new
     config.active_record.sqlite3.represent_boolean_as_integer = nil
 
+    config.active_record.queues = ActiveSupport::InheritableOptions.new(destroy: :active_record_destroy)
```
I guess it depends on what we want to expose as the public API. E.g. just the `ActiveRecord::DestroyJob` and the others to inherit from, or a succinct setting for many queues.

For the latter, something like this is now possible:

```ruby
# config/initializers/queues.rb
Rails.application.config.active_record.queues[:destroy] = …
Rails.application.config.active_storage.queues[:purge] = …
```

rafaelfranca replied:
We will have to end up exposing both because even that we ship with a job to destroy record later I already see a problem where your application requires a more complex job than the one we ship. Shopify for sure will need a custom job because of our infra.

georgeclaghorn commented on activerecord/lib/active_record/destroy_job.rb:
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class DestroyJob < ActiveJob::Base
```
Yeah, I’d prefer to raise if no destroy job is configured.

gwincr11 replied:
I think I agree with @georgeclaghorn, raise and emit a message that this is not supported without the inclusion of ActiveJob? I think it is misleading to complete a behavior that is not available in a different way, and if you have a noisy app you may never see a warning message. Essentially it would be like progressive enhancement in a web browser, you can still use active record but not every feature is available to you unless you bring in ActiveJob.

rafaelfranca replied:
I'm ok with that as well. It just that it doesn't make sense to raise if we can actually complete the work. I see this more as the inline adapter to Active Job than as progressive enhancement.

kaspth commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,21 @@
+      destroy_later after: 30.days, if: -> { completed_previously_changed? && completed? }
+
+      # Destroy exports 30 days after completing them, ensuring they're still completed at destroy time
+      destroy_later after: 30.days, if: -> { completed_previously_changed? && completed? }, ensuring: :completed?
```
Just ported this to an app I'm working on and noticed that `completed_previously_changed?` should be `status_previously_changed?` (or whatever the underlying column is called). 😊

gwincr11 commented on activerecord/lib/active_record.rb:
```diff
@@ -26,6 +26,7 @@
 require "active_support"
 require "active_support/rails"
 require "active_model"
+require "active_job"
```
Can we conditionally include this? https://www.rubydoc.info/github/rubygems/rubygems/Gem%2FSpecification.find_by_name

```suggestion
begin
    Gem::Specification.find_by_name "active_job"
    require "active_job"
    rescue Gem::LoadError
       # maybe log out a warning?
    end
```

Then in future calls we can say
```ruby
do_thing if defined?(ActiveJob)
```

Just a thought...

rafaelfranca replied:
We can but this is not we decouple things in Rails. The way I explained it in https://github.com/rails/rails/pull/34901#discussion_r24647389 is the way we do.

rafaelfranca commented:
Closing this PR since we decided to not implement the class level `destroy_later`. `dependent: :destroy_async` was implemented in #40157.
